### PR TITLE
Add a base64 ARCH_CMD encoder

### PR DESCRIPTION
**Pull request by zeroSteiner**
This adds a new encoder module that leverages base64 encoding to escape bad characters in ARCH_CMD payloads for the Linux and UNIX platforms. It aims to consolidate the way in which modules have been doing it by automatically identifying the decoding binary, or allowing the user to specify it themselves.

## Testing

- [ ] Use an ARCH_CMD payload module in msfconsole
- [ ] Generate the payload with one or more bad characters that will trigger the encoding
- [ ] See that the encoder works and the payload executes

### CODE REVIEW

File: modules/encoders/cmd/base64.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    when 'base64'
+      base64_decoder = 'base64 -d'
+    when 'openssl'
+      base64_decoder = 'openssl enc -base64 -d'
+    else
+      # find a decoder at runtime if we can use the necessary characters
+      if (state.badchars.bytes & '>()/|-&'.bytes).empty?
+        base64_decoder = '(which base64 >/dev/null && base64 -d || which openssl >/dev/null && openssl enc -base64 -d)'
+      else
+        base64_decoder = 'base64 -d'
```
Comment (bcoles) [inline modules/encoders/cmd/base64.rb:57]: If we want to support OSX, `--decode` would be a safer fallback default flag.

File: lib/msf/core/encoder.rb [hunk 2]
```diff
@@ -142,6 +142,10 @@ module Type
     # Bash brace expansion encoding.
     #
     CmdPosixBrace = 'brace'
+    #
+    # Bas64 encoding.
```
Comment (bcoles) [inline lib/msf/core/encoder.rb:146]:
```suggestion
    # Base64 encoding.
```

File: modules/encoders/cmd/base64.rb (hunk 1 above)
Comment (bcoles) [reply modules/encoders/cmd/base64.rb:57]: Apparently BusyBox `base64` is the opposite: `-d` is supported but `--decode` is not.

https://boxmatrix.info/wiki/Property:base64_(bbcmd)

File: modules/encoders/cmd/base64.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      ],
+      self.class
+    )
+  end
+
+  #
+  # Encodes the payload
+  #
+  def encode_block(state, buf)
+    return buf if state.badchars.empty?
```
Comment (jvoisin) [inline modules/encoders/cmd/base64.rb:38]: Shouldn't `buf` be returned as well if it doesn't contain any bytes present in `state.badchars`?

File: modules/encoders/cmd/base64.rb (hunk 1 above)
Comment (jvoisin) [reply modules/encoders/cmd/base64.rb:57]: For busybox, `busybox base64 -d` can be used to tell it apart from coreutils' busybox.

File: modules/encoders/cmd/base64.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    base64_buf = Base64.strict_encode64(buf)
+    case datastore['Base64Decoder']
+    when 'base64'
+      base64_decoder = 'base64 -d'
+    when 'openssl'
+      base64_decoder = 'openssl enc -base64 -d'
+    else
+      # find a decoder at runtime if we can use the necessary characters
+      if (state.badchars.bytes & '>()/|-&'.bytes).empty?
+        base64_decoder = '(which base64 >/dev/null && base64 -d || which openssl >/dev/null && openssl enc -base64 -d)'
```
Comment (jvoisin) [inline modules/encoders/cmd/base64.rb:55]: `which` [isn't portable](https://stackoverflow.com/questions/592620/how-can-i-check-if-a-program-exists-from-a-bash-script), `command -v` is ~~better~~ POSIX, so it should work even on weird platforms.

File: modules/encoders/cmd/base64.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    base64_buf = Base64.strict_encode64(buf)
+    case datastore['Base64Decoder']
+    when 'base64'
+      base64_decoder = 'base64 -d'
+    when 'openssl'
+      base64_decoder = 'openssl enc -base64 -d'
+    else
+      # find a decoder at runtime if we can use the necessary characters
+      if (state.badchars.bytes & '>()/|-&'.bytes).empty?
```
Comment (jvoisin) [inline modules/encoders/cmd/base64.rb:54]: Wouldn't something like this be more clear/maintainable:

```ruby
base64_decoder = '(which base64 >/dev/null && base64 -d || which openssl >/dev/null && openssl enc -base64 -d)'
if (state.badchars.bytes & base64_decode.to_a.uniq.to_s.bytes).empty?
  base64_decoder = 'base64 -d'
end
```

File: modules/encoders/cmd/base64.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      else
+        base64_decoder = 'base64 -d'
+      end
+    end
+
+    if (state.badchars.bytes & '-|'.bytes).empty?
+      buf = "echo #{base64_buf}|#{base64_decoder}|sh"
+    elsif (state.badchars.bytes & '-<()'.bytes).empty?
+      buf = "sh < <(#{base64_decoder} < <(echo #{base64_buf}))"
+    else
```
Comment (jvoisin) [inline modules/encoders/cmd/base64.rb:65]: Speaking of horrible golfing:

```suggestion
    elsif (state.badchars.bytes & '<`-'.bytes).empty?
      buf = "sh<<<`#{base64_decoder}<<<'#{base64_buf}'`
    else
```

**Review by cdelafuente-r7 — APPROVED** [review]

**Comment by cdelafuente-r7** [comment]
Thanks @zeroSteiner ! It looks good to me. I tested every generated payloads on Linux (Debian) and Mac OS X and it works as expected. I'll go ahead and land it.

**Pull request closed without merging**